src/managedwriter/writer_client.ts

```diff
@@ -0,0 +1,311 @@
+ * @class
+ * @memberof storage
+ */
+
+type StreamConnection = {
+  write_stream: WriteStream;
+  connection: gax.CancellableStream;
+};
+type streamConnectionsMap = Record<string, gax.CancellableStream>;
+type StreamConnections = {
+  connection_list: StreamConnection[];
+  connections: streamConnectionsMap;
+};
+type WriteStream = protos.google.cloud.bigquery.storage.v1.IWriteStream;
+type AppendRowResponse =
+  protos.google.cloud.bigquery.storage.v1.AppendRowsResponse;
+type AppendRowRequest =
+  protos.google.cloud.bigquery.storage.v1.IAppendRowsRequest;
+type IInt64Value = protos.google.protobuf.IInt64Value;
+type ProtoData =
+  protos.google.cloud.bigquery.storage.v1.AppendRowsRequest.IProtoData;
+
+export class WriterClient {
+  private _opts: ClientOptions | undefined;
+  private _parent: string;
+  private _writeStreamType: WriteStream['type'] = 'TYPE_UNSPECIFIED';
+  private _streamId: string;
+  private _client: BigQueryWriteClient;
+  private _connections: StreamConnections;
+  private _client_closed: boolean;
+
+  constructor(
+    parent?: string,
+    client?: BigQueryWriteClient,
+    bqWriteClientOpts?: ClientOptions,
+    writeStreamType?: WriteStream['type']
+  ) {
+    this._parent = parent ? parent : 'Please set a parent path';
+    this._client = new BigQueryWriteClient(bqWriteClientOpts) || client;
+    this._writeStreamType = writeStreamType || this._writeStreamType;
+    this._connections = {
+      connection_list: [],
+      connections: {},
+    };
+    this._streamId = 'Please open a connection to set connection name';
+    this._client_closed = false;
+  }
+
+  setParent = (projectId: string, datasetId: string, tableId: string): void => {
+    const parent = `projects/${projectId}/datasets/${datasetId}/tables/${tableId}`;
+    this._parent = parent;
+  };
+
+  getParent = (): string => {
+    return this._parent;
+  };
+
+  getStreamId = (): string => {
+    return this._streamId;
+  };
+
+  setStreamId = (streamId: string): void => {
+    this._streamId = streamId;
+  };
+
+  getClient = (): BigQueryWriteClient => {
+    return this._client;
+  };
+
+  setClient = (client: BigQueryWriteClient): void => {
+    this._client = client;
+  };
+
+  getWriteStreams = (writeStream: WriteStream): undefined | null | string[] => {
+    if (writeStream === undefined || writeStream.name === undefined) {
+      return undefined;
+    }
+    if (writeStream.name === null) {
+      return null;
+    }
+    return new Array(writeStream.name);
+  };
+
+  setWriteStreamType(streamType: WriteStream['type']): void {
+    this._writeStreamType = streamType;
+  }
+
+  getWriteStreamType(): WriteStream['type'] {
+    return this._writeStreamType;
+  }
+
+  getConnections(): StreamConnections {
+    return this._connections;
+  }
+
+  getClientClosedStatus(): boolean {
+    return this._client_closed;
+  }
+
+  async initializeStreamConnection(clientOptions?: CallOptions): Promise<void> {
```

Comment:
I think we need to split this method into two actions: Create a WriteStream ( if applicable, as users can use the [DefaultStream](https://cloud.google.com/bigquery/docs/write-api-streaming#at-least-once) and create a "ManagedStream", which is a concept of stream connection open to append rows to it.

Reply:
The `ManagedStream` idea came from the Go implementation. The Python implementation has a similar concept called StreamSession. That class will have a reference to a given streamId, schema/proto descriptor and will have a connection opened to append rows. That connection can be reused.